src/audio-worklet/free-queue/src/interface/free_queue.h

```diff
@@ -0,0 +1,233 @@
+
+#ifdef FREE_QUEUE_IMPL
+
+static uint32_t _getAvailableRead(
+  struct FreeQueue *queue, 
+  uint32_t readIndex, 
+  uint32_t writeIndex
+) {  
+  if (writeIndex >= readIndex)
+    return writeIndex - readIndex;
+  
+  return writeIndex + queue->buffer_length - readIndex;
+}
+
+static uint32_t _getAvailableWrite(
+  struct FreeQueue *queue, 
+  uint32_t readIndex, 
+  uint32_t writeIndex
+) {
+  if (writeIndex >= readIndex)
+    return queue->buffer_length - writeIndex + readIndex - 1;
+  
+  return readIndex - writeIndex - 1;
+}
+
+struct FreeQueue *CreateFreeQueue(size_t length, size_t channel_count) {
+  struct FreeQueue *queue = (struct FreeQueue *)malloc(sizeof(struct FreeQueue));
+  queue->buffer_length = length + 1;
+  queue->channel_count = channel_count;
+  queue->state = (atomic_uint *)malloc(2 * sizeof(atomic_uint));
+  atomic_store(queue->state + READ, 0);
+  atomic_store(queue->state + WRITE, 0);
+
+  queue->channel_data = (float **)malloc(channel_count * sizeof(float *));
+  for (int i = 0; i < channel_count; i++) {
+    queue->channel_data[i] = (float *)malloc(queue->buffer_length * sizeof(float));
+    for (int j = 0; j < queue->buffer_length; j++) {
+      queue->channel_data[i][j] = 0;
+    }
+  }
+  return queue;
+}
+
+void DestroyFreeQueue(struct FreeQueue *queue) {
+  for (int i = 0; i < queue->channel_count; i++) {
+    free(queue->channel_data[i]);
+  }
+  free(queue->channel_data);
+  free(queue);
+}
+
+bool FreeQueuePush(struct FreeQueue *queue, float **input, size_t blockLength) {
+  uint32_t currentRead = atomic_load(queue->state + READ);
+  uint32_t currentWrite = atomic_load(queue->state + WRITE);
+
+  if (_getAvailableWrite(queue, currentRead, currentWrite) < blockLength) {
+    return false;
+  }
+
+  for (uint32_t i = 0; i < blockLength; i++) {
+    for (uint32_t channel = 0; channel < queue->channel_count; channel++) {
+      queue->channel_data[channel][(currentWrite + i) % queue->buffer_length] = 
+          input[channel][i];
+    }
+  }
+
+  uint32_t nextWrite = (currentWrite + blockLength) % queue->buffer_length;
+  atomic_store(queue->state + WRITE, nextWrite);
+  return true;
+}
+
+bool FreeQueuePull(struct FreeQueue *queue, float **output, size_t blockLength) {
+  uint32_t currentRead = atomic_load(queue->state + READ);
+  uint32_t currentWrite = atomic_load(queue->state + WRITE);
+
+  if (_getAvailableRead(queue, currentRead, currentWrite) < blockLength) {
+    return false;
+  }
+
+  for (uint32_t i = 0; i < blockLength; i++) {
+    for (uint32_t channel = 0; channel < queue->channel_count; channel++) {
+      output[channel][i] = 
+          queue->channel_data[channel][(currentRead + i) % queue->buffer_length];
+    }
+  }
+
+  uint32_t nextRead = (currentRead + blockLength) % queue->buffer_length;
+  atomic_store(queue->state + READ, nextRead);
+  return true;
+}
+
+void *GetFreeQueuePointers(struct FreeQueue *queue, char *data) {
+  if (strcmp(data, "buffer_length") == 0) {
```

Comment:
Extracting this "buffer_length" string into a dictionary/enum would be better. Same for other struct members below.

Reply:
I think in C it's not possible to use dictionary and enum too for string comparison.
Though I will look into that, if I can have more abstraction here in this part